File: lintcode/sliding-window-maximum.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/sliding-window-maximum/

class Solution:
    """
    @param nums: A list of integers.
    @return: The maximum number inside the window at each moving.
    """
    def maxSlidingWindow(self, nums, k):
        # write your code here
        if len(nums) == 0 or len(nums) < k:
            return []

        deque = []
        for i in xrange(k):
            while deque and nums[i] >= nums[deque[-1]]:
                deque.pop()
            deque.append(i)

        res = []
        for i in xrange(k, len(nums)):
            res.append(nums[deque[0]])
            while deque and nums[i] >= nums[deque[-1]]:
                deque.pop()
            while deque and deque[0] <= i-k:
                deque.pop(0)
            deque.append(i)

        res.append(nums[deque[0]])
        return res

if __name__ == '__main__':
    s = Solution()
    print s.maxSlidingWindow([1, 2, 7, 7, 8], 3)

File: leetcode-2/sliding-window-maximum.cc
// Question: https://leetcode.com/problems/sliding-window-maximum/

class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> ans;
        if (nums.empty()) {
            return ans;
        }
        deque<int> q;
        for (int i = 0; i < k; ++i) {
            while (!q.empty() && nums[q.back()] < nums[i]) {
                q.pop_back();
            }
            q.push_back(i);
        }

        for (int i = k; i < nums.size(); ++i) {
            ans.push_back(nums[q.front()]);
            while (!q.empty() && q.front() <= i-k) {
                q.pop_front();
            }
            while (!q.empty() && nums[q.back()] < nums[i]) {
                q.pop_back();
            }
            q.push_back(i);
        }

        ans.push_back(nums[q.front()]);

        return ans;
    }
};
